lib/dry/system/auto_registrar.rb

```diff
@@ -28,12 +29,14 @@ def finalize!
 
       # @api private
       def call(dir, &block)
+        config = AutoRegistraionConfig.new
```

Comment:
For this part of the code I had different solutions in mind I wanted to share with you to see which one fits more with your idea. I will try to explain what the other idea I had.

1.- The one in the PR, we leave the `config` instance the logic to decide if the proc object for `instance` or `exclude` has been store, one thing that I do not like is that we always instanciate `AutoRegistraionConfig ` even when there is no block passed

2.- The other was this one 
```ruby
def call(dir, &block)
   yield(config = AutoRegistraionConfig.new) if block_given?
   components(dir).each do |component|
      next if !component.auto_register? || (config && config.call(:exclude, component))

  container.require_component(component) do
    if initialized_component = config && config.call(:instance, component)
      register(component.identifier) { initialized_component }
    else
      register(component.identifier) { component.instance }
    end
  end
 end
```

This solution avoid creating unneeded instance of `config` but since there is chance `config` is `nil` we have to double check, not really fun of this one

3.- The other solution would instantiate the configuration in the initializer of `auto_register`

```ruby
def initialize(container)
  @container = container
  @config = container.config
  @auto_registration_config = AutoRegistraionConfig.new
end

# @api private
def call(dir, &block)
  yield(auto_registration_config) if block_given?
  components(dir).each do |component|
    next if !component.auto_register? || auto_registration_config.call(:exclude, component)

    container.require_component(component) do
       if initialized_component = auto_registration_config.call(:instance, component)
         register(component.identifier) { initialized_component }
       else
         register(component.identifier) { component.instance }
       end
    end
  end
end
```